### ISSUE TITLE
windows support

### ISSUE THREAD

**Issue by perrywang** [rancher/k3s#114]
there are pretty much customers are using windows at edge side device with installation of labview.

**Comment by ibuildthecloud** [comment]
This one is tricky.  I personally don't know much about Kubernetes on Windows, but Rancher/RKE works on windows so I can find out what needs to be done here.  If there's demand, I'd love to do this.

**Comment by perrywang** [comment]
on windows in general it is using a vm hyperv or vbox to run kubenetes like minikube doing. But minikube is really not production ready and there is a lot of issue and no any customization capability.

**Comment by bencompton** [comment]
Looking at Microsoft's instructions for [joining a Windows node to a cluster](https://docs.microsoft.com/en-us/virtualization/windowscontainers/kubernetes/joining-windows-workers?tabs=ManagementIP), Kubernetes releases have [binaries](https://github.com/kubernetes/kubernetes/blob/master/CHANGELOG-1.13.md#node-binaries) for running Windows worker nodes (kubelet.exe, kube-proxy.exe), but not the control plane. I see there's also an official [Windows binary for Flannel](https://github.com/coreos/flannel/releases), which Microsoft's PS scripts leverage.

I would think a K3s Windows build with only the agent components would be possible. At this point, it looks like [Containerd support](https://docs.microsoft.com/en-us/virtualization/windowscontainers/deploy-containers/containerd#containerdcri) in Windows is still only in alpha and the official Windows build is [still a WIP](https://github.com/containerd/cri/issues/560). Not sure what other dependencies in K3s would be problematic for an agent-only Windows build.

**Comment by mortezaalizadeh** [comment]
Hey guys, I just checked the Containerd support for Windows and sounds like it's no longer in alpha state. I am using K3S two of my side projects and supporting Windows on the edge is one of my highest priorities in both projects.

We also don't really need to run the control plane on the edge and Agent support on the Window should be more than enough. I do not know enough about the implementation details of the K3S and K8S, but I am more than happy to help to get the Windows support done.

Is there any progress on this ticket since the last comment posted on 11 April?

**Comment by KnicKnic** [comment]
my fork can start and join a windows master node. It does appear to have done a few things right, example my kubelet lists container images on my node (I am using docker desktop community). However I don't have networking setup, so the node stays unschedulable.

I would love for someone with more understanding of network configuration to take it further. And give instructions on how to setup networking. @ibuildthecloud do you know anyone that could help here?

cmd I used to get it to "somewhat" startup
```
.\k3s.exe server -d c:\tmp\k3s --flannel-backend none --docker  --kube-proxy-arg "proxy-mode=userspace" --disable-network-policy
```

Branch I was using for this experiment. https://github.com/KnicKnic/k3s/tree/k3s_run_windows_node


There are few errors (these related to cni / networking)
```
W0223 00:59:39.172343   20244 cni.go:237] Unable to update cni config: no networks found in /etc/cni/net.d
E0223 00:59:39.937342   20244 kubelet.go:2183] Container runtime network not ready: NetworkReady=false reason:NetworkPluginNotReady message:docker: network plugin is not ready: cni config uninitialized
```
And this random one for TLS, have no clue why a cert would be wrong.
```
2020-02-23 00:54:37.591274 I | http: TLS handshake error from 127.0.0.1:2111: remote error: tls: bad certificate
2020-02-23 00:54:37.602363 I | http: TLS handshake error from 127.0.0.1:2114: remote error: tls: bad certificate
```

**Comment by KnicKnic** [comment]
@thxCode @alena1108 @cloudnautique @aiwantaozi pinging some other people in rancher that seemed to have worked on windows integration (for rancher) to see if they could help with setting up windows networking.

**Comment by teyc** [comment]
@KnicKnic I just bookmarked this https://github.com/microsoft/SDN/blob/master/Kubernetes/flannel/register-svc.ps1 a bunch of script for establishing a SDN on Windows.

**Comment by jiridanek** [comment]
Even though I can't run k3s agent on Windows right now, at this point, it should be possible to connect a k8s node to a k3s control node, yes?

**Comment by brandond** [comment]
@jiridanek vanilla k8s kubelets cannot connect to k3s servers, as they are missing the k3s-specific bootstrap code.

**Comment by maximveksler** [comment]
> @jiridanek vanilla k8s kubelets cannot connect to k3s servers, as they are missing the k3s-specific bootstrap code.

Any way to work around that? Perhaps a local adapter of some sort?

**Comment by andyxhadji** [comment]
Very interested in a k3s windows agent as well, has there been any updates here? Happy to contribute here as well

**Comment by gruselglatz** [comment]
@ibuildthecloud  Is there some progress or status on the Windows K3s Agent? Is there some Branch or something to work on to speed things up?

Thanks

**Comment by miketzian** [comment]
+1 would be keen for this (windows agent + linux server), what is the "k3s-specific bootstrap code" ? (if there's an arch doc, I can look at that)

**Comment by wtfacoconut** [comment]
+1 I too would like to have a package to add in a windows host as a worker node!  :)

**Comment by moley** [comment]
+1 We'd like to use k3s in a windows environment too to handle linux pods

**Comment by viceice** [comment]
Yeah, joining windows worker nodes would be enough for our use case ❤️

**Comment by brandond** [comment]
Windows nodes will run Windows containers only. This is true Kubernetes on Windows, not Kubernetes in WSL or anything like that.

**Comment by viceice** [comment]
Yes, that's what I need. I've some Linux VM for running k3s and want to join some windows VM for running workload which only works on Windows.

**Comment by moley** [comment]
> > @moley
> > +1 We'd like to use k3s in a windows environment too to handle linux pods
> 
> Windows nodes will run Windows containers only. This is true Kubernetes on Windows, not Kubernetes in WSL or anything like that.

Ok, so we will need a linux vm in our windows environment to get this running? Are there plans to enable this directly in the future?

**Comment by lippertmarkus** [comment]
Now with Docker Desktop being a paid solution for enterprises this would be valuable for Rancher Desktop.

**Comment by brandond** [comment]
Rancher Desktop already runs K3s on Windows without any additional work necessary.

**Comment by lippertmarkus** [comment]
But no Windows Containers which this issue is about, right?

**Comment by wtfacoconut** [comment]
Hello all. I just wanted to check. Has anyone been able to get a windows worker node to join a k3s cluster?

**Comment by andruwa13** [comment]
In Nomad (https://www.nomadproject.io) the network is normally working on a plugin https://github.com/containernetworking/plugins

**Comment by wtfacoconut** [comment]
> We do not currently have any plans to add native Windows support to K3s.

@brandond , pleeeeeease? 

![image](https://user-images.githubusercontent.com/5297545/134751296-3540d7ca-b3ae-4de4-8d99-8428836c6410.png)

**Comment by pablos91** [comment]
> Hello. Maybe a different question to ask. But would it be possible to manually add a windows node to a K3s cluster? Has anyone tried? If so, did you encounter any issues?

You could probably experiment with rancher desktop on Windows BUT it will only run Linux containers via WSL2 so it is probably not what everyone is wishing for - a windows container runtime inside a cluster.

**Comment by bangzhuzhu** [comment]
@wtfacoconut have you tried it?

**Comment by wtfacoconut** [comment]
> @wtfacoconut have you tried it?
> 

Are you referring to the recommendation to use WSL2 on Windows? If so then the answer is no, because this would still only run Linux containers.

**Comment by bangzhuzhu** [comment]
>@bangzhuzhu, are you referring to the recommendation to use WSL2 on Windows? If so then the answer is no, because this would still only run Linux containers.

I mean manually add a windows node(k8s work node) to a k3s cluster,

**Comment by bangzhuzhu** [comment]
> vanilla k8s kubelets cannot connect to k3s servers, as they are missing the k3s-specific bootstrap code.

@brandond Would mind telling something more about it? Thank you

**Comment by tmyroadctfig** [comment]
For anyone else for bumps into this thread, Microsoft have started working on AKS Edge, which has an option for k3s and for Windows worker nodes.
https://learn.microsoft.com/en-us/azure/aks/hybrid/aks-edge-overview
https://github.com/Azure/AKS-Edge

The k3s version of AKS Edge currently spins up a linux hyper-v VM with k3s on it

**Comment by brandond** [comment]
@tmyroadctfig yes, we're aware. The source for AKS Edge's fork is at https://github.com/seanyen/k3s/tree/dev/seanyen/windows. There are ongoing efforts to get this upstreamed.

**Comment by viceice** [comment]
Windows agent seems to basically work ❤️ 
- #7258 🎉

Now there need to be a prebuild k3s.exe on releases so this can be tested more widely. 🙃

**Comment by salvioli** [comment]
My understanding is that this is not yet supported, is this correct? What is the current status? What are the missing steps to provide windows support?

**Comment by caroline-suse-rancher** [comment]
There was a hope that we could improve our Windows support, and we're now resource constrained and unable to devote time to this. I will close this as currently not planned. If that changes in the future, or we get community contributions, we can reopen.
